code/modules/wiremod/port.dm

```diff
@@ -52,185 +53,165 @@
+			return new_type == PORT_TYPE_STRING || new_type == PORT_TYPE_SIGNAL
 		if(PORT_TYPE_SIGNAL)
 			// A signal port is just a number port but distinguishable
-			return datatype_to_check == PORT_TYPE_NUMBER
+			return new_type == PORT_TYPE_NUMBER
 
 	return FALSE
 
-/**
- * # Input Port
- *
- * An input port that can only be connected to 1 output port
- *
- * Registers a signal on the target output port to listen out for any output
- * so that an update can be sent to the attached component
- */
+/datum/port/output
 /datum/port/input
-	/// The output value of the port
-	var/input_value
-
-	/// The connected output port
-	var/datum/port/output/connected_port
-
 	/// Whether this port triggers an update whenever an output is received.
 	var/trigger = FALSE
 
-	/// The default value of this input
-	var/default
+	/// The ports this port is wired to.
+	var/list/datum/port/connected_ports
 
 /datum/port/input/New(obj/item/circuit_component/to_connect, name, datatype, trigger, default)
 	. = ..()
+	put(default)
 	src.trigger = trigger
-	src.default = default
-	set_input(default, FALSE)
+	src.connected_ports = list()
 
 /**
- * Connects the input port to the output port
- *
- * Sets the input_value and registers a signal to receive future updates.
- * Arguments:
- * * port_to_register - The port to connect the input port to
+ * Introduces two ports to one another.
  */
-/datum/port/input/proc/register_output_port(datum/port/output/port_to_register)
-	unregister_output_port()
-
-	RegisterSignal(port_to_register, COMSIG_PORT_SET_OUTPUT, .proc/receive_output)
-	RegisterSignal(port_to_register, list(
-		COMSIG_PORT_DISCONNECT,
-		COMSIG_PARENT_QDELETING
-	), .proc/unregister_output_port)
-
-	connected_port = port_to_register
-	SEND_SIGNAL(connected_port, COMSIG_PORT_OUTPUT_CONNECT, src)
+/datum/port/input/proc/connect(datum/port/output/output)
+	SIGNAL_HANDLER
+	if(!compatible_datatypes(output.datatype, src.datatype))
+		return
+	connected_ports |= output
+	RegisterSignal(output, COMSIG_WIREMOD, .proc/receive)
 	// For signals, we don't update the input to prevent sending a signal when connecting ports.
 	if(datatype != PORT_TYPE_SIGNAL)
-		set_input(connected_port.output_value)
-
+		put(output.value)
 
 /**
- * Sets a timer depending on the value of the input_receive_delay
- *
- * The timer will call a proc that updates the value.
- * Arguments:
- * * connected_port - The connected output port
- * * new_value - The new value received from the output port
+ * Handle an output port update.
+ * Values are handled with an input_receive_delay
  */
-/datum/port/input/proc/receive_output(datum/port/output/connected_port, new_value)
+/datum/port/input/proc/receive(datum/port/output/output, new_value=null, new_type=null, disconnect=FALSE)
 	SIGNAL_HANDLER
-	SScircuit_component.add_callback(CALLBACK(src, .proc/set_input, new_value))
+	if(disconnect)
+		disconnect(output)
+	else if(new_type)
+		if(!compatible_datatypes(new_type, src.datatype))
+			disconnect(output)
```

Comment:
remove this check, the entire point is that you shouldn't need to do this check

Reply:
I have more than once been annoyed when I changed the type of a router and had to reconnect the wires. I don't see the safety.

Reply:
sure it's not a big deal, it's just easy to get wrong